README.md
# Fallback Site for Craft 3

This plugin allows each site to fall back to another site when a requested element cannot be found. This allows multi-site installations (typically for multiple languages) to gracefully fall back to existing content if available.

When a request is made for an element that would normally not be found, *Fallback Site* will check the configured site for an element with the same slug. For example, if you have an English site and a Spanish site as part of your craft install, and English is the fallback for Spanish, then handling for a request to the Spanish site would go as follows:

* First, if an element was found on the Spanish site, everything continues as normal, and *Fallback Site* does not intervene.
* If an element was not found on the Spanish site, *Fallback Site* checks if there is an English element with the same the slug.
* If an English element is found, then that element's ID is used to see if there is a Spanish version available, which can happen if an incorrect, untranslated slug is used.
	* If a Spanish version exists, then a 301 redirect will be sent to the user, directing them to the correct URL.
	* If a Spanish version does not exist, then *Fallback Site* makes Craft use the English element in place of the Spanish one (With some conditions listed below).
* If an English element is not found, then the request continues as a 404 as normal.

See the flowcharts below which demonstrate what happens when requesting a page, `about-us`, under a secondary, Spanish site (located at `/es`) within a multi-site Craft installation.

**Normal Craft element handling**

![Craft routing without Fallback Site](./resources/without-plugin.png)

**Element handling with Fallback Site**

![Craft routing with Fallback Site](./resources/with-plugin.png)

## Requirements

* Craft CMS 3.0.0-RC1 or above

##
1. Open a terminal and navigate to your project folder:

```bash
cd /path/to/project
```

2. Require the package to download it via Composer:

```bash
composer require charliedev/fallback-site
```

3. Install the plugin on the `Settings -> Plugins` page in the Craft control panel.

**--- or ---**

3. Install the plugin via the command line:

```bash
./craft install/plugin fallback-site
```

## Configuration

![Fallback Site settings panel](./resources/settings-screenshot.png)

Within the Craft control panel, go to the `Settings -> Plugins` page, and then look for the `Settings` button listed by the *Fallback Site* plugin.

Here, you can select which site acts as a fallback for each other site. Fallbacks can also be chained across multiple sites, each site falling back to the next.

## Usage

While the plugin works without any additional configuration, you may want to know if an entry being displayed has been substituted by *Fallback Site*. This may be done to display notices about the availability of site-specific content, or notify users of the replacement. Detecting this can be done by comparing the current site stored by Craft with the entry's associated site.

```
{% if craft.app.sites.currentSite != entry.site %}
	{# Entry content is coming from a different site. #}
{% endif %}
{# ... or ... #}
{% if craft.app.language != entry.site.language %}
	{# Entry content is in a different language. #}
{% endif %}
```

---

*Built for [Craft CMS](https://craftcms.com/) by [Charlie Development](http://charliedev.com/)*


CHANGELOG.md
# Changelog

## Unreleased

## 1.2.2 - 2019-04-16

### Fixed
- Now calls `Craft::$app->end()` instead of simply halting the script, to allow Craft the opportunity to run any cleanup. (Thanks @stevepsharpe).

## 1.2.1 - 2018-11-05

### Fixed
- Initialization moved from post plugin load (`Plugins::EVENT_AFTER_LOAD_PLUGINS`) to post application initialization (`Application::EVENT_INIT`).

## 1.2.0 - 2018-02-26

### Fixed
- Fallback Site now performs initialization after all plugins have loaded, its old behavior could cause compatibility issues with other plugins.
- Set correct `license` option in the `composer.json`.

## 1.1.0 - 2018-02-26

### Added
- Readme now has a screenshot of the settings panel.

### Fixed
- Altered fallback behavior to not replace Craft's UrlManager component, which was interfering with other routing plugins, like Element API.

## 1.0.0 - 2018-01-19

The initial release of the Fallback Site plugin.

### Added
- Requests for elements that don't exist on a site, but exist under a fallback site, are substituted in place.
- Each site can have a specific fallback site configured for it.

src/FallbackSite.php
<?php
/**
 * Fallback Site plugin for Craft 3.0
 * @copyright Copyright Charlie Development
 */

namespace charliedev\fallbacksite;

use Craft;
use craft\base\Plugin;
use craft\web\Application;
use craft\web\UrlManager;

use yii\base\Event;

/**
 * The main Craft plugin class.
 */
class FallbackSite extends Plugin
{
	/**
	 * @inheritdoc
	 * @see craft\base\Plugin
	 */
	public function init()
	{
		parent::init();

		// NOTE: This will wind up using Craft's `UrlManager`.
		// This means any plugins/modules that have not yet attached their event handlers to UrlManager will never
		// receive some of its events (EVENT_REGISTER_CP_URL_RULES, EVENT_REGISTER_SITE_URL_RULES). By this point
		// plugins should have added their events, but any that use the UrlManager improperly might cause trouble
		// because of the early initialization.
		Event::on(
			Application::class,
			Application::EVENT_INIT,
			function () {
				$this->checkFallbackEntry();
			}
		);
	}

	/**
	 * @inheritdoc
	 * @see craft\base\Plugin
	 */
	protected function createSettingsModel()
	{
		return new \charliedev\fallbacksite\models\Settings();
	}

	/**
	 * @inheritdoc
	 * @see craft\base\Plugin
	 */
	protected function settingsHtml()
	{
		$sites = Craft::$app->getSites()->getAllSites();
		$siteoptions = []; // A list of options, tailored to each site.

		foreach ($sites as $site) {
			$siteoptions[$site->id] = []; // Build a list for the given site.
			$siteoptions[$site->id][$site->id] = Craft::t('app', 'None'); // A site using its own ID does not fallback, and is a default 'none' option,
			foreach ($sites as $option) { // Build a list of all other sites.
				if ($option->id != $site->id) { // Do not re-add the same site to the list.
					$siteoptions[$site->id][$option->id] = Craft::t('site', $option->name);
				}
			}
		}

		return Craft::$app->getView()->renderTemplate('fallback-site/_settings', [
			'settings' => $this->getSettings(),
			'siteoptions' => $siteoptions
		]);
	}

	/**
	 * Attempts to find and render an entry from configured fallback sites if an entry hasn't been found for the visited site.
	 */
	private function checkFallbackEntry() {

		// Only run for regular web frontend requests.
		if (!Craft::$app->getRequest()->getIsConsoleRequest()
		&& Craft::$app->getRequest()->getIsSiteRequest()
		&& !Craft::$app->getRequest()->getIsLivePreview()) {
			// Determine if there is an element at the given URL, and attempt to find one using fallback sites
			// if one is not available. This may interfere with custom routes or direct templates, if one
			// happens to conflict with the potential path.
			$element = Craft::$app->getUrlManager()->getMatchedElement(); // Find the element that the current URL is going to.
			if ($element == null) { // No element found, this request may 404 normally.
				$this->renderFallbackEntry();
			}
		}
	}

	/**
	 * Attempts to find and render an entry from configured fallback sites if an entry hasn't been found for the visited site.
	 */
	private function renderFallbackEntry()
	{
		$originalsite = Craft::$app->getSites()->currentSite; // Store the original site being visited.
		$siteid = $originalsite->id; // Storage for the ID of the site currently being checked.
		$path = Craft::$app->getRequest()->getPathInfo(); // Retrieve the path being visited, used for checking for entries.
		$checked = [$originalsite->id => true]; // Mark the current site as checked.
		$sites = $this->getSettings()->sites; // Retrieve the set of sites and their fallbacks.

		// Check the fallback for each site until a valid entry is found, or the fallback is a site that has already been checked .
		while (true) {
			if (!isset($sites[$siteid])) { // No fallback (bad config, unconfigured site, etc.).
				return;
			}
			$siteid = $sites[$siteid]; // Get the ID of the fallback site for the last checked site.
			if (isset($checked[$siteid])) { // Fallback site has already checked (avoid cyclical references), give up.
				return;
			}
			$checked[$siteid] = true; // Mark this new site as checked.

			$element = Craft::$app->getElements()->getElementByUri($path, $siteid, true); // Check for an entry with the same path in the fallback site.
			if ($element) { // An element was found with the given path and site id.
				// Make sure the element has a route, too.
				$route = $element->getRoute();
				if ($route) {
					// Check to see if the element has an enabled version with the original site, in case it actually just has a different slug.
					$originalelement = Craft::$app->getElements()->getElementById($element->id, null, $originalsite->id);
					if ($originalelement && $originalelement->getStatus() == 'live') { // Found an element with a different slug that is available.
						Craft::$app->getResponse()->redirect($originalelement->getUrl(), 301); // Redirect to the proper element.
						Craft::$app->end(); // Redirect immediately, do not let Craft fall back to default 404 behavior (which wipes out the redirect).
					} else {
						// Do some reflection **magic** to substitute the element and route with the cached ones in Craft's UrlManager.
						// This is required because there is no other way to clear this cache short of replacing the UrlManager component
						// which results in a loss of all attached events and associated route parameters.
						// If Craft changes these properties, this will break. Ideally the UrlManager would provide means of resetting this internal state.
						$reflector = new \ReflectionClass(\craft\web\UrlManager::class);
						$elementproperty = $reflector->getProperty('_matchedElement');
						$elementproperty->setAccessible(true);
						$elementproperty->setValue(Craft::$app->getUrlManager(), $element);
						$routeproperty = $reflector->getProperty('_matchedElementRoute');
						$routeproperty->setAccessible(true);
						$routeproperty->setValue(Craft::$app->getUrlManager(), $route);
					}
				}
				return; // Found an entry, possibly redirected or rerouted (or one with no route). Work here is done, continue as normal.
			}
		}
	}
}


src/models/Settings.php
<?php
/**
 * Fallback Site plugin for Craft 3.0
 * @copyright Copyright Charlie Development
 */

namespace charliedev\fallbacksite\models;

use Craft;
use craft\base\Model;

/**
 * The plugin settings model for the Fallback Site plugin.
 */
class Settings extends Model
{

	public $sites = [];

	/**
	 * @inheritdoc
	 * @see yii\base\BaseObject
	 */
	public function init()
	{

		$sites = Craft::$app->getSites()->getAllSiteIds();

		// Each site falls back to itself alone.
		foreach ($sites as $site) {
			$this->sites[$site] = $site;
		}

		parent::init();
	}

	/**
	 * @inheritdoc
	 * @see craft\base\Model
	 */
	public function rules()
	{
		return [
			['sites', 'validateSites']
		];
	}

	/**
	 * Makes sure that the sites and their configured fallbacks are valid.
	 */
	public function validateSites(string $attribute)
	{
		if (!is_array($this->sites)) {
			$this->addError($attribute, Craft::t('fallback-site', 'Sites must be provided as an array'));
			return;
		}
		$sites = Craft::$app->getSites()->getAllSiteIds();
		foreach ($this->sites as $site => $fallback) {
			if (!in_array($site, $sites)) {
				$this->addError($attribute, Craft::t('fallback-site', '"' . $site . '" is not a valid site ID'));
			}
			if (!in_array($site, $sites)) {
				$this->addError($attribute, Craft::t('fallback-site', '"' . $fallback . '" is not a valid site ID'));
			}
		}
	}
}


src/DefaultSite.php
<?php
/**
 * Default Site plugin for Craft 3.0
 * @copyright Copyright Charlie Development
 */

namespace charliedev\defaultsite;

use Craft;
use craft\base\Plugin;
use craft\events\ExceptionEvent;
use craft\helpers\ArrayHelper;
use craft\services\Fields;
use craft\web\ErrorHandler;
use craft\web\UrlManager;

use yii\base\Event;
use yii\web\HttpException;

/**
 * The main Craft plugin class.
 */
class DefaultSite extends Plugin
{
	/**
	 * @inheritdoc
	 * @see craft\base\Plugin
	 */
	public function init()
	{
		// Ideally it would be nice to do something where, between entry checking and dynamic routes,
		// the lack of an entry is caught and able to be optionally handled/replaced.
		// Because this isn't elegantly possible, another solution is to catch 404 errors and render
		// output if possible then, but this runs the risk of interfering/competing with other plugins
		// that may also perform actions on 404 pages, and also comes with the requirement of having
		// to force ending the script in order to avoid discarding the rendered output, which means
		// Yii/Craft never have the opportunity to finish what they're doing.
		/*
		Event::on(
			ErrorHandler::class,
			ErrorHandler::EVENT_BEFORE_HANDLE_EXCEPTION,
			function (ExceptionEvent $event) {
				// Don't do special 404 handling for console requests or previews.
				if (Craft::$app->getRequest()->getIsConsoleRequest() || !Craft::$app->getRequest()->getIsSiteRequest() || Craft::$app->getRequest()->getIsLivePreview()) {
					return;
				}

				$exception = $event->exception;
				
				if ($exception instanceof \Twig_Error_Runtime
					&& $exception->getPrevious() !== null) {
                    $exception = $exception->getPrevious();
                }
				
				if ($exception instanceof HttpException
					&& $exception->statusCode === 404) {
					$this->renderFallbackEntry();
                }
			},
			false
		);
		*/

		// This section, instead of going the event-based approach, will preemptively determine if
		// there is an element at the given URL, and attempt to find one if not, using as much of
		// Craft's built-in functionality as possible. The big difference between this and the
		// above code, is that it could potentially block requests from running custom routes and
		// direct templates, if an entry on a fallback site exists that conflicts with that path.
		// In addition, because it hooks into Craft's own system of rendering, it may be less likely
		// to interfere with other plugins or behaviors that would normally run for entry rendering
		// or on 404 handling.
		if (!Craft::$app->getRequest()->getIsConsoleRequest() && Craft::$app->getRequest()->getIsSiteRequest() && !Craft::$app->getRequest()->getIsLivePreview()) { // Only run for regular web frontend requests.
			$element = Craft::$app->getUrlManager()->getMatchedElement(); // Find the element that the current URL is going to.
			if ($element == null) { // No element found, this request will 404 normally.
				$originalsite = Craft::$app->getSites()->currentSite; // Get the current site, for reference.
				$checked = [$originalsite->id => true]; // Mark the current site as checked.
				$sites = $this->getSettings()->sites;
				$currentsite = Craft::$app->getSites()->currentSite->id;

				// Check the fallback for each site until a valid entry is found, or the fallback is a site that has already been checked .
				while (true) {
					if (!isset($sites[$currentsite])) { // No fallback (bad config, unconfigured site, etc.).
						return;
					}
					$currentsite = $sites[$currentsite];
					if (isset($checked[$currentsite])) { // Already checked this site, give up.
						return;
					}
					$checked[$currentsite] = true; // Mark this new site as checked.
					Craft::$app->getSites()->currentSite = Craft::$app->getSites()->getSiteById($currentsite); // Force Craft into new site.
					Craft::$app->set('urlManager', new UrlManager()); // Reset the URL Manager component, otherwise the request can't properly be re-processed under the new site.
					Craft::$app->getUrlManager()->parseRequest(Craft::$app->getRequest()); // Re-parse the request to find a potential new route.
					$element = Craft::$app->getUrlManager()->getMatchedElement(); // Find the element matched with the fallback site, if possible.
					Craft::$app->getSites()->currentSite = $originalsite; // Revert Craft to the original site, in case the element has been found.
					
					if ($element != false) { // If an element with the fallback site was found, check if it has an enabled version with the original site, in case of different slugs.
						$element = Craft::$app->getElements()->getElementById($element->id); // Try to retrieve the element with the same ID, using the original locale.
						if ($element && $element->getStatus() == 'live') { // The element exists in the original site and is enabled, redirect to it.
							Craft::$app->getRequest()->redirect($element->getUrl(), true, 301); // Element was found under original site, but with different slug.
						}
						return; // Element was found, but has no enabled version on the original site, fallback to it instead.
					}
				}
			}
		}

		parent::init();
	}

	/**
	 * @inheritdoc
	 * @see craft\base\Plugin
	 */
	protected function createSettingsModel()
    {
        return new \charliedev\defaultsite\models\Settings();
	}
	
	/**
	 * @inheritdoc
	 * @see craft\base\Plugin
	 */
	protected function settingsHtml()
    {
		$sites = Craft::$app->getSites()->getAllSites();
		$siteoptions = [];

		foreach ($sites as $site) {
			$siteoptions[$site->id] = Craft::t('site', $site->name);
		}
        return Craft::$app->getView()->renderTemplate('default-site/_settings', [
			'settings' => $this->getSettings(),
			'sites' => $siteoptions
        ]);
	}
	
	/**
	 * Attempts to find and render an entry from configured fallback sites if an entry hasn't been found for the visited site.
	 */
	private function renderFallbackEntry() {
		$originalsite = Craft::$app->getSites()->currentSite; // Store the original site being visited.
		$siteid = $originalsite->id; // Storage for the ID of the site currently being checked.
		$path = Craft::$app->getRequest()->getPathInfo(); // Retrieve the path being visited, used for checking for entries.
		$checked = [$originalsite->id => true]; // Mark the current site as checked.
		$sites = $this->getSettings()->sites; // Retrieve the set of sites and their fallbacks.

		// Check the fallback for each site until a valid entry is found, or the fallback is a site that has already been checked .
		while (true) {
			if (!isset($sites[$siteid])) { // No fallback (bad config, unconfigured site, etc.).
				return false; // Default back to normal 404 behavior.
			}
			$siteid = $sites[$siteid]; // Get the ID of the fallback site for the last checked site.
			if (isset($checked[$siteid])) { // Fallback site has already checked (avoid cyclical references), give up.
				return false; // Default back to normal 404 behavior.
			}
			
			$element = Craft::$app->getElements()->getElementByUri($path, $siteid, true); // Check for an entry with the same path in the fallback site.
			if ($element) { // An element was found with the given path and site id.
				
				// Make sure the element has a route, too.
				$route = $element->getRoute();
				if ($route) {

					// Check to see if the element has an enabled version with the original site, in case it actually just has a different slug.
					$originalelement = Craft::$app->getElements()->getElementById($element->id, null, $originalsite->id);
					if ($originalelement && $originalelement->getStatus() == 'live') { // Found an element with a different slug.
						Craft::$app->controller->redirect($originalelement->getUrl(), true, 301); // Redirect to the proper element.
						die(); // End here, so the output isn't cleared and we don't get a 404 response.
					} else {

						/** @see \yii\web\Request::resolve() */
						$route[1] += $_GET; // Additional parameters should be attached to match what Yii already does internally.

						$render = Craft::$app->runAction($route[0], $route[1]); // Render the template with the updated entry.
						$render->send(); // Send the rendered template to the user.
						die(); // End here, so the output isn't cleared and we don't get a 404 response.
					}
				}
				return false; // Found an entry, but it has no route? Probably not an entry meant to be visited, continue with 404.
			}
		}
	}
}
